src/components/Form/FormField.tsx

```diff
@@ -0,0 +1,221 @@
+    className: PropTypes.string,
+
+    /**
+     * A form control component (i.e. Input) or HTML tagName (i.e. 'input').
+     * Extra FormField props are passed to the control component.
+     * Mutually exclusive with children.
+     */
+    controlType: customPropTypes.some([
+      PropTypes.func,
+      PropTypes.oneOf(['button', 'input', 'select', 'textarea']),
+    ]),
+
+    /** A control for the form field. */
+    control: customPropTypes.itemShorthand,
+
+    /**
+     * A custom render function for the control slot.
+     *
+     * @param {React.ReactType} Component - The computed component for this slot.
+     * @param {object} props - The computed props for this slot.
+     * @param {ReactNode|ReactNodeArray} children - The computed children for this slot.
+     */
+    renderControl: PropTypes.func,
+
+    /** A label for the form field. */
+    label: customPropTypes.itemShorthand,
+
+    /**
+     * A custom render function for the label slot.
+     *
+     * @param {React.ReactType} Component - The computed component for this slot.
+     * @param {object} props - The computed props for this slot.
+     * @param {ReactNode|ReactNodeArray} children - The computed children for this slot.
+     */
+    renderLabel: PropTypes.func,
+
+    /** A field can have its label next to instead of above it. */
+    inline: PropTypes.bool,
+
+    /** A field can show that input is mandatory. */
+    required: PropTypes.bool,
+
+    /** Text message that will be displayed below the control (can be used for error, warning, success messages). */
+    message: customPropTypes.itemShorthand,
+
+    /**
+     * A custom render function for the message slot.
+     *
+     * @param {React.ReactType} Component - The computed component for this slot.
+     * @param {object} props - The computed props for this slot.
+     * @param {ReactNode|ReactNodeArray} children - The computed children for this slot.
+     */
+    renderMessage: PropTypes.func,
+
+    /** Additional CSS styles to apply to the component instance.  */
+    styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
+
+    /** Override for theme site variables to allow modifications of component styling via themes. */
+    variables: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
+  }
+
+  public static defaultProps = {
+    as: 'div',
+  }
+
+  public renderComponent({
+    ElementType,
+    classes,
+    accessibility,
+    variables,
+    styles,
+    rest,
+  }): React.ReactNode {
+    const {
+      children,
+      control,
+      renderControl,
+      label,
+      renderLabel,
+      id,
+      type,
+      required,
+      message,
+      renderMessage,
+      controlType,
+    } = this.props
+
+    let labelContent = label
+
+    if (required) {
+      labelContent =
+        typeof label === 'string'
+          ? `${label} *`
+          : typeof label === 'object' && (label as any).content
+            ? { ...label, content: `${(label as any).content} *` }
```

Comment:
I don't think we should make so many presumptions on the content of the `label` shorthand; this looks like a candidate for a smaller component that will handle all this logic

Reply:
I think I over-complicated things here :\ I will just add additional text component after the label if the field is required, without changing the actual text inside the label.